### PR TITLE
Refactor lonlat_to_cartesian to header-only API

### PR DESCRIPTION
**Pull request by harrism**
Following #477, implements the header-only API for `cuspatial::lonlat_to_cartesian` and implements existing C++ API on top of it.

Follows the refactoring guide introduced in #477.

Note this branch is based on the branch for #477 so diff includes all changes from that PR as well. Once #477 is merged this PR will simplify a lot.

### CODE REVIEW

**Review by vyasr — APPROVED** [review]
CMake LGTM

File: cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  Location _origin{};
+};
+
+}  // namespace detail
+
+template <class InputIt, class OutputIt, class Location>
+OutputIt lonlat_to_cartesian(InputIt lon_lat_first,
+                             InputIt lon_lat_last,
+                             OutputIt xy_first,
+                             Location origin,
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh:74]: If `Location` can only be `lonlat_2d<T>`, why not just make it `lonlat_2d<T>`?

Comment (harrism) [reply cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh]: Good point. Fixed.

File: cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                             lonlat_2d<T> origin,
+                             rmm::cuda_stream_view stream)
+{
+  static_assert(std::is_floating_point_v<T>,
+                "lonlat_to_cartesian supports only floating-point coordinates.");
+
+  CUSPATIAL_EXPECTS(origin.x >= -180 && origin.x <= 180 && origin.y >= -90 && origin.y <= 90,
+                    "origin must have valid longitude [-180, 180] and latitude [-90, 90]");
+
+  return thrust::transform(rmm::exec_policy(stream),
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh:83]: I was wondering, for methods that does not fit into the `transform` framework (e.g. linestring instance that launches on `num_points` and writes to `num_segments` output), what's the STL convention to follow? cc @jrhemstad

Comment (jrhemstad) [reply cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh:83]: What exactly at your looking for? An example of an algorithm that operates on N inputs and produces M outputs where N != M? 

Stream compaction algorithms like `copy_if` are one example ( M <= N).

Offhand I can't think of algorithms where M >= N. Maybe `set_union`? 

@codereport would be the real guru to ask here.

Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh:83]: For pairwise linestring distance the input the data structure is segmented, and each pair of segments (non-geometric meaning) map to a single value (distance). It's like segmented reduction with two value input iterators.

Comment (codereport) [reply cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh:83]: @isVoid Do you mean two elements at a time are mapped to a single value?  Something like this:
```hs
Prelude> [1..10]                        -- [1,2,3,4,5,6,7,8,9,10]
Prelude> chunksOf 2 [1..10]             -- [[1,2],[3,4],[5,6],[7,8],[9,10]]
Prelude> map sum . chunksOf 2 $ [1..10] -- [3,7,11,15,19]
```

Comment (harrism) [reply cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh:83]: @codereport no. Two separate segmented input arrays. Equal number of segments in the two arrays, but corresponding segments need not have the same length.  Apply an operator to corresponding pairs of segments.

File: cpp/include/cuspatial/experimental/coordinate_transform.cuh [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ * @param[in]  stream: The CUDA stream on which to perform computations and allocate memory.
+ *
+ * All input iterators must have a `value_type` of `cuspatial::lonlat_2d<T>` (Lat/Lon coordinates),
+ * and the output iterator must be able to accept for storage values of type
+ * `cuspatial::cartesian_2d<T>` (Cartesian coordinates).
+ *
+ * @tparam InputIt Iterator over longitude/latitude locations. Must meet the requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @tparam OutputIt Iterator over Cartesian output points. Must meet the requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/coordinate_transform.cuh:43]: I noticed that for output iterators we should also mention it's **mutable**.

https://en.cppreference.com/w/cpp/named_req/OutputIterator

**Review by isVoid — APPROVED** [review]
I think it looks good. Just one small question above.

File: cpp/include/cuspatial/experimental/coordinate_transform.cuh (hunk 3 above)
Comment (harrism) [reply cpp/include/cuspatial/experimental/coordinate_transform.cuh]: Done. Thanks!

**Comment by harrism** [comment]
rerun tests

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**